system/Database/BaseResult.php

```diff
@@ -130,7 +123,7 @@ public function getCustomResultObject(string $className)
 			return $this->customResultObject[$className];
 		}
 
-		if (is_bool($this->resultID) || ! $this->resultID || $this->numRows === 0)
+		if (is_bool($this->resultID) || ! $this->resultID || $this->getNumRows() === 0)
```

Comment:
This is a really bad idea. When there is no native support for getting a number of rows for the query (like SQLite3), we will experience a significant performance hit for every query - even if we don't need to use `getNumRows()` in our own code.

Reply:
> For the `SQLSVR` problem, I'm not sure why it's not working now but the flag we're using is valid per documentation. The flag is inherited from CI3 and it seems a good fit.

I think the code in [`Database\BaseUtils::optimizeTable`](https://github.com/sneakyimp/CodeIgniter4/blob/develop/system/Database/BaseUtils.php#L126) is problematic. I see that the `OPTIMIZE TABLE` query in MySQLi does in fact return a result that can produce a record set, but it seems weird to be calling getResultArray on an optimize or alter query because that function assumes the query result contains some iterable record set. Furthermore, this function is only ever called from tests/system/Database/Live/DbUtilsTest and the results [there](https://github.com/sneakyimp/CodeIgniter4/blob/develop/tests/system/Database/Live/DbUtilsTest.php#L126) just cast the output as boolean. Seems to me that `BaseUtils::optimizeTable` function should return `true` on success, `false` on failure. That said, it's hard to tell what the intent is with many of these functions because the return types and behaviors in the javadoc comments are not very specific. A return type of *mixed* or *object* is not nearly as helpful as *boolean|null* or *Database\ResultInterface*.

I would be happy to add an optimizeTable method to the various db Util classes so that their peculiarities can be distilled into a simple boolean value. However, I wonder if that might be better off handled in a different pull request?

> As you mentioned there are different flags for `SQLSRV` so there is also a question should we support the `getNumRows()` for all of them or not (with some of them `sqlsrv_num_rows()` will not work).

Some queries do not return result rows. For example [mysqli_query](https://www.php.net/manual/en/mysqli.query.php) returns booleans in some cases:

>Returns false on failure. For successful SELECT, SHOW, DESCRIBE or EXPLAIN queries mysqli_query() will return a mysqli_result object. For other successful queries mysqli_query() will return true.

> This is a really bad idea. When there is no native support for getting a number of rows for the query (like SQLite3), we will experience a significant performance hit for every query - even if we don't need to use `getNumRows()` in our own code.

@michalsn, you make an excellent point, which I admit I had overlooked. These getResultArray functions etc. could be called over and over so the getNumRows function would have to be very efficient indeed. I don't know if you looked at the latest code I've committed, but I did implement the caching behavior you suggested (using a numRows property relocated specifically to the SQLite3\Result class). It is quite fast -- even faster than relying on the native mysqli_num_rows function. I wrote this test in a controller:

```php
	public function foo() {
		$db = \Config\Database::connect("tests");
		echo $db->getDatabase() . "\n";
		
		$query = $db->query("select * from db_job");
		
		// measure the first call (which requires loop thru records
		$start = microtime(TRUE);
		$v1 = $query->getNumRows();
		$first = microtime(TRUE) - $start;
		echo "first: $first\n";
		
		// measure the cumulative/avg time for 1000 interations
		$cumulative = 0;
		for($i=0; $i<1000; $i++) {
			$start = microtime(TRUE);
			$v2 = $query->getNumRows();
			$cumulative += (microtime(TRUE) - $start);
			if ($v1 != $v2) {
				throw new \Exception("$v1 is not equal to $v2");
			}
		}
		echo "cumulative: $cumulative\n";
		echo "average: " . ($cumulative/1000) . "\n";
		$query->freeResult();

	}
```
I tested it for SQLite3 with predictable and encouraging results:
```
db1.db
first: 2.5033950805664E-5
cumulative: 0.0001070499420166
average: 1.070499420166E-7
```
the first result requires us to iterate the result set to get the count so it's much slower, but subsequent requests are consistently very fast.

Compare that result to the mysqli code, which just calls mysqli_num_rows every time. I ran that a few times and the speeds vary:
```
ci4_test
first: 5.0067901611328E-6
cumulative: 0.00033020973205566
average: 3.3020973205566E-7

ci4_test
first: 2.8610229492188E-6
cumulative: 0.00012063980102539
average: 1.2063980102539E-7

ci4_test
first: 3.0994415283203E-6
cumulative: 0.00019383430480957
average: 1.9383430480957E-7
```
I think a getNumRows function running in 100-200 nanoseconds seems sufficiently fast.

HOWEVER, I think we may continue to have problems if we call getResultArray on any old query result. As you point out, some query results do not have any records and sqlsrv_num_rows returns false to signal an error condition. I admit I'm a bit unsure how to proceed. I think the getNumRows method will be plenty fast to run in an oft-used function like getResultArray, but I worry that folks may be calling that on query results which were never intended to yield records.

Reply:
I may sound a bit rude here (sorry for that), but I don't care how fast it will be - iterating the whole result to count the number of rows for every `SELECT` query is just wrong. Especially that there are other possibilities.

You don't know if anybody will ever need this value, so why waste the time and resources to do that? Simply pointless.

I don't believe I will change my mind on this one. `getNumRows()` should work on demand and always, because in another case it will be a source of frustration for developers. 

Fun fact - personally I don't think we need this method at all. Because it's pretty much useful only when you're using `getUnbufferedRow()`. In all other cases simply using `count()` is good enough. So, I guess this was one of the reasons this method wasn't ported to CI4 in the first place.

Reply:
> I may sound a bit rude here (sorry for that), but I don't care how fast it will be - iterating the whole result to count the number of rows for every `SELECT` query is just wrong. Especially that there are other possibilities.

I don't mind rude -- and this may sound a bit rude in return -- but I don't think you are paying attention to what the current code is doing or to my latest commit. The develop branch right now without my changes checks a numRows property in getResultArray, getResultObject, and getCustomResultObject *that is never, ever set in the first place*. To be clear, the original developer of these functions was themself attempting to use numRows value *for the sake of efficiency*.

I would also point out that in my proposed solution, only *one* of the four supported db drivers, SQLite3, would ever iterate through all the records to count them. All of the other db drivers offer a num_rows function. Furthermore, I have argued here that, if you want to know whether your query has returned a result, the only other alternatives I have seen proposed here or on the CodeIgniter forum are to use Builder->countAllResults or to **iterate the entire result and instantiate every record.**

> You don't know if anybody will ever need this value, so why waste the time and resources to do that? Simply pointless.

As I pointed out [two days ago](https://github.com/codeigniter4/CodeIgniter4/pull/4049#issuecomment-753528132) my CI3 project that I am currently migrating uses the old query->num_rows function in no less than a hundred places.

> I don't believe I will change my mind on this one. `getNumRows()` should work on demand and always, because in another case it will be a source of frustration for developers.

Have you not read the code? Have you not tested it? It does always work in my latest commit to this pull request. In the case of SQLite3, it must unfortunately iterate the entire result set exactly *once* the first time it is called but on any subsequent calls, no recordset iteration is necessary because the first numRows value is cached. If getNumRows is *not* called, this would never happen and the variable Result::numRows variable remains NULL. In the case of SQLSRV, it also works. The reason that DBUtilsTest is failing is due to a badly written test, which I have pointed out in issue #4059.

And, as long as we are discussing our feelings about poor coding decisions, I would, at the very least, remove the current numRows checks in getResultArray, getResultObject, and getCustomResultObject functions as they currently exist in the develop branch because they are checking a class property that never even gets set in the first place. I would also examine BaseResult::query because it appears to instantiate a ResultInterface object even if the query failed and resultID is `false`. This is also mentioned in #4059 but might deserve its own issue report.

> Fun fact - personally I don't think we need this method at all. Because it's pretty much useful only when you're using `getUnbufferedRow()`. In all other cases simply using `count()` is good enough. So, I guess this was one of the reasons this method wasn't ported to CI4 in the first place.

I regret that we disagree. Please note that using count() implies that you have *iterated the entire result set and instantiated an array or object from each record.* If you are using very large result sets, or *even if you are just checking to see if it's large before you instantiate all those objects* that is hardly a satisfactory option. The CI3 project I am migrating to CI4 refers to the unbuffered row function in 123 places and handles result sets with thousands of records. This is precisely what has prompted me to attempt this pull request in the first place.

Reply:
> I don't mind rude -- and this may sound a bit rude in return -- but I don't think you are paying attention to what the current code is doing or to my latest commit. The develop branch right now without my changes checks a numRows property in getResultArray, getResultObject, and getCustomResultObject that is never, ever set in the first place. To be clear, the original developer of these functions was themself attempting to use numRows value for the sake of efficiency.

When I'm commenting on code you may be sure I'm paying attention to the changes that were made.

I'm aware that `$numRows` was never set (as I pointed it out in my second comment in this discussion). But bear in mind that this small change in code has certain consequences. `$numRows` would be `null` as default, but using `getNumRows()` is forcing a particular action. 

> I would also point out that in my proposed solution, only one of the four supported db drivers, SQLite3, would ever iterate through all the records to count them. All of the other db drivers offer a num_rows function. Furthermore, I have argued here that, if you want to know whether your query has returned a result, the only other alternatives I have seen proposed here or on the CodeIgniter forum are to use Builder->countAllResults or to iterate the entire result and instantiate every record.

It doesn't matter if it affects only one driver. Making an automatic call to this method should not happen. It should happen only if I, as a developer will make an explicit call in my code to the `getNumRows()` method.

> As I pointed out two days ago my CI3 project that I am currently migrating uses the old query->num_rows function in no less than a hundred places.

I would ask you if you really need it.

> Have you not read the code? Have you not tested it? It does always work in my latest commit to this pull request.

It's not true. Try to use `freeResult()` and then use `getNumRows()`. And you don't have to explain to me that I should not do this. I saw enough of the code to know that this was a very common thing - in the context of migrating from CI3.

> If getNumRows is not called, this would never happen and the variable Result::numRows variable remains NULL.

This is the main problem. The thing is it is called every single time when I want to iterate over the results. Please check one more time under which method we're having this discussion.

> And, as long as we are discussing our feelings about poor coding decisions, I would, at the very least, remove the current numRows checks in getResultArray, getResultObject, and getCustomResultObject functions as they currently exist in the develop branch because they are checking a class property that never even gets set in the first place.

Compared to the current proposal, I agree that it would be a better solution.

> I regret that we disagree. Please note that using count() implies that you have iterated the entire result set and instantiated an array or object from each record. 

I'm perfectly aware of it. The thing is that we usually check the number of the rows, and if there are any, we iterate over them anyway.

> The CI3 project I am migrating to CI4 refers to the unbuffered row function in 123 places and handles result sets with thousands of records. This is precisely what has prompted me to attempt this pull request in the first place.

If you really use `getUnbufferedRow()` in 123 places then I guess you would really benefit from having this method available. I have no other questions.

Reply:
>It doesn't matter if it affects only one driver. Making an automatic call to this method should not happen. It should happen only if I, as a developer will make an explicit call in my code to the getNumRows() method.

I agree 💯

>I would ask you if you really need it.

My need for it is based on two concerns. The first concern is that we are dealing with (admittedly unusual) cases where a) the original query is very painful due to fulltext search of large amounts of data and b) it is not clear how many records, if any, might be returned. To instantiate a very large record set has in the past caused fatal `out of memory` errors -- we must check before trying to instantiate. To run the query again, even with a COUNT(*) instead of the original select, is also painful due to the large amounts of data searched, REGEXP or LIKE expressions, etc. The query cache will not help. You will note that, even if the query cache would come into play, it has been [removed in MySQL 8](https://dev.mysql.com/doc/refman/5.7/en/query-cache-status-and-maintenance.html). The second concern is to migrate an existing CI3 codebase to CI4. This will be a lot harder if we must examine over 200 spots in our code to assess how we might otherwise solve the issue. There's a very real cost to the development effort that might require.

I'd also point out that MySQLi, Postgres, and SQL Server all offer a num_rows function. Surely their wisdom reflects a real need?

>It's not true. Try to use freeResult() and then use getNumRows(). And you don't have to explain to me that I should not do this. I saw enough of the code to know that this was a very common thing - in the context of migrating from CI3.

I believe that any db driver will return false/null/exception if you try to call free_result on a query result after you have called free_result on it.  For example:
```php
$sql = "SELECT * FROM db_job";
$params = [];
$options =  array( "Scrollable" => SQLSRV_CURSOR_KEYSET );
$stmt = sqlsrv_query( $conn, $sql , $params, $options );
if( $stmt === false ) {
     die( print_r( sqlsrv_errors(), true));
}
echo "num_rows before: " . sqlsrv_num_rows($stmt) . "\n";
sqlsrv_free_stmt($stmt);
$aft = sqlsrv_num_rows($stmt);
echo "num_rows after: " . sqlsrv_num_rows($stmt) . "\n";
var_dump($aft);
```
`$aft` in this case is `false` and the code emits an E_WARNING.

In MySQLi:
```php
if ($result = $mysqli->query("SELECT * from db_job")) {

    /* determine number of rows result set */
    $row_cnt = $result->num_rows;
    printf("Result set has %d rows.\n", $row_cnt);

    $result->free_result();
    $aft = $result->num_rows;
    var_dump($aft);

    /* close result set */
    $result->close();
}
```
$aft in this case is NULL and any reference to $result after free_result emits an E_WARNING "Couldn't fetch mysqli_result".

I must ask for some examples to illustrate what you mean here. Until someone accepts this pull request, there is no getNumRows function in CodeIgniter 4 so we have an opportunity to define behavior as we see fit. I propose that calling `Result->getNumRows()` after you have called `Result->freeResult` should either throw an exception or return zero/false/null because the resultID will have been completely wiped. If you are referring to CI3 behavior with query->num_rows() then I suggest that no one should be calling num_rows after calling free_result and recommend we take this major version to behave more like the underlying raw database functions. If you think the functionality should be different, I'd be happy to hear your suggestion.

>This is the main problem. The thing is it is called every single time when I want to iterate over the results. Please check one more time under which method we're having this discussion.
>Compared to the current proposal, I agree that it would be a better solution.

I will remove the checks to numRows in getResultArray, getResultObject, and getCustomResultObject and commit again.

>I'm perfectly aware of it. The thing is that we usually check the number of the rows, and if there are any, we iterate over them anyway.

I cannot find any check performed anywhere in the current CodeIgniter 4 development branch where the number of rows returned by a query is checked before they are iterated. I find no reference to the string "num_rows" anywhere in the development branch system folder except for a comment in system/Database/SQLSRV/Connection.php:
```
$ grep -r "num_rows" system
system/Database/SQLSRV/Connection.php:	 * but would disable num_rows() (and possibly insert_id())
```

Reply:
> I believe that any db driver will return false/null/exception if you try to call free_result on a query result after you have called free_result on it. 

The point is that I'm referring to `num_rows()` method from CI3, which as you know acts differently. And I believe that if we want to make this function usable it should offer the same functionality. I persistently return to this argument because it is crucial here if we are talking about migrating apps from CI3.

> I must ask for some examples to illustrate what you mean here. Until someone accepts this pull request, there is no getNumRows function in CodeIgniter 4 so we have an opportunity to define behavior as we see fit. I propose that calling Result->getNumRows() after you have called Result->freeResult should either throw an exception or return zero/false/null because the resultID will have been completely wiped.

I would stay with my initial idea which is quite similar to one that was implemented in CI3. We should simply have a "failover" which will use `count()` on results, that's all. This way we make sure that this method can be used every time without worrying about the current `resultID` status.

> I cannot find any check performed anywhere in the current CodeIgniter 4 development branch where the number of rows returned by a query is checked before they are iterated. I find no reference to the string "num_rows" anywhere in the development branch system folder except for a comment in system/Database/SQLSRV/Connection.php:

I guess we haven't understood each other correctly. I was referring to the way developers usually use this function when they write their own code. So, that's why I wrote in my previous post that this function is handy only for usage with `getUnbufferedRow()` method and in all other cases using `count()` is good enough. 

That being said my personal preference would be to not have `getNumRows()` method at all since it might be confusing (under some circumstances with current implementation) and not implemented in every driver in native manner. In my projects, when I need `num_rows`, I just make a clear call: `$db->table('whatever')->get()->resultID->num_rows` and that's enough. It would look a little different for other drivers but I guess you know what I mean. Anyway, it's not really relevant since we want to have this method implemented.

Reply:
> The point is that I'm referring to `num_rows()` method from CI3, which as you know acts differently. And I believe that if we want to make this function usable it should offer the same functionality. I persistently return to this argument because it is crucial here if we are talking about migrating apps from CI3.

I maintain that calling num_rows after free_result is a mistake and that any reasonable developer would/should avoid doing that. If you truly believe this is common practice in CI3 projects, I'd be very interested in seeing some examples. I've never done it myself. Could you tell me *why* you think we should continue to show a non-empty numRows value after the result has been freed? To me this is a bit like trying to refer to a variable after it has been `unset()` and should be discouraged.

You are correct that the CI3 [mysqli](https://github.com/bcit-ci/CodeIgniter/blob/18718545e0faa98d62df3f65d045d84494ec29ca/system/database/drivers/mysqli/mysqli_result.php#L58), [PostGreSQL](https://github.com/bcit-ci/CodeIgniter/blob/18718545e0faa98d62df3f65d045d84494ec29ca/system/database/drivers/postgre/postgre_result.php#L58), and [SQLSRV](https://github.com/bcit-ci/CodeIgniter/blob/18718545e0faa98d62df3f65d045d84494ec29ca/system/database/drivers/sqlsrv/sqlsrv_result.php#L80) drivers do first attempt to use a cached num_rows value and fall back on checking the count of resultArray or resultObjects that have been fetched. Interestingly, the sqlsrv driver checks for a scrollable cursor and if the cursor is inappropriate, falls back on the parent class definition, which only checks any resultArray/resultObject that might have been fetched. I tend to agree that if we want to preserve some continuity that we might want to imitate this prior behavior, but am somewhat conflicted.

> I would stay with my initial idea which is quite similar to one that was implemented in CI3. We should simply have a "failover" which will use `count()` on results, that's all. This way we make sure that this method can be used every time without worrying about the current `resultID` status.

Falling back onto counting resultArray and resultObject does make sense. It looks to me like CI3 [first attempts](https://github.com/bcit-ci/CodeIgniter/blob/develop/system/database/DB_result.php#L130) to return a previously fetched row_num property (if any). Failing that, it returns count(resultArray). If there is no fetched result set, it will fetch the result set and count it.

> I guess we haven't understood each other correctly. I was referring to the way developers usually use this function when they write their own code. So, that's why I wrote in my previous post that this function is handy only for usage with `getUnbufferedRow()` method and in all other cases using `count()` is good enough.

getNumRows could still be useful if you are worried that the size of the query result set might be too large before you call getResultArray. This might be true whether you use getUnbufferedRow or not. 

> That being said my personal preference would be to not have `getNumRows()` method at all since it might be confusing (under some circumstances with current implementation) and not implemented in every driver in native manner. In my projects, when I need `num_rows`, I just make a clear call: `$db->table('whatever')->get()->resultID->num_rows` and that's enough. It would look a little different for other drivers but I guess you know what I mean. Anyway, it's not really relevant since we want to have this method implemented.

I think you've made a good case to mimic prior CI3 behavior, and the decisions made there look pretty wise to me. As for this:
```php
$rowCount = $db->table('whatever')->get()->resultID->num_rows;
```
That may work for MySQLi but it won't for any other dbms, to it is not portable.

I'd like to make another commit that mimics the CI3 behavior, but there's a bit of an issue that BaseConnection::query is frequently constructing ResultInterface objects where resultID is a boolean value (false even!). For this getNumRows function, I will first check the resultID of a given ResultInterface object and, if the resultID is boolean (or a non-scrollable cursor type). If the resultID is boolean, this seems like undefined behavior. Two options:
1) Throwing an exception seems like it would be useful to developers because it would discourage careless attempts to count non-countable queries and encourage error checking for failed queries.
2) return NULL and leave $this->numRows NULL; Is there any CI4 best practice for emitting a warning?

Reply:
> I maintain that calling num_rows after free_result is a mistake and that any reasonable developer would/should avoid doing that. If you truly believe this is common practice in CI3 projects, I'd be very interested in seeing some examples. I've never done it myself. Could you tell me why you think we should continue to show a non-empty numRows value after the result has been freed?

I think the "problem" is that you subconsciously combine the `getNumRows()` method with the equivalent from the `mysqli_num_rows()`, `pg_num_rows()` etc. For the developer, it is simply a method which name suggests returning the number of records that were returned by the query. I've seen code that looks like this many times:
```php
$query = $this->db->get('whatever')
if ($query->num_rows())
{
	$results = $query->result_array()
	$query->free_results();
	...
}
...
if ($query->num_rows())
{
	// make another db call
}
```
or many different combinations. The name of this method just suggests returning the number of rows, so expecting returning this value every time is nothing surprising. People in general use frameworks to not be bothered by details - the method should just do what it promises. So if we take away this functionality at some point it would be frustrating. And it's not that we don't know how many rows are there now (after using `free_results()`). We still have this information, so why not make it available? Especially if there is a simple workaround - like in this case.

> That may work for MySQLi but it won't for any other dbms, so it is not portable.

That's true, I mentioned I would require a different approach for other drivers.

> I'd like to make another commit that mimics the CI3 behavior, but there's a bit of an issue that BaseConnection::query is frequently constructing ResultInterface objects where resultID is a boolean value (false even!).

1. if debug mode is enabled (`$this->DBDebug`) then we should throw an exception. Regardless of the status for debug mode we should always use `log_message('error', 'error message')` for logging the error
2. if debug mode is not enabled, then we should return `0` because I feel like returning numeric value makes more sense here. Internally when`$numRows` is `null`, he knows that `getNumRows()` were never called.

Reply:
Thank you so much for your continued input!

> I think the "problem" is that you subconsciously combine the `getNumRows()` method with the equivalent from the `mysqli_num_rows()`, `pg_num_rows()` etc. For the developer, it is simply a method which name suggests returning the number of records that were returned by the query. I've seen code that looks like this many times:

Thank you very much for the code example. I've never seen CodeIgniter code that would extract a query's results, free the result object, and then refer to the freed query result to get a record count when you already have a variable containing those results. I would also point out that the point of calling free_result is to free up memory. Looking at CI3, I see that query->free_result does nothing to free up the arrays/objects that might have been fetched from the query, which surprises me. I would indeed prefer to mimic the behavior of mysqli_num_rows, etc., because I believe that is the behavior most developers would expect, IMHO, a proper freeResult method would also clear any resultArray or resultObject arrays inside the query object. On the other hand, migration concerns and backwards compatibility matter. One can always do `unset($query)`, right?

> or many different combinations. The name of this method just suggests returning the number of rows, so expecting returning this value every time is nothing surprising. People in general use frameworks to not be bothered by details - the method should just do what it promises. So if we take away this functionality at some point it would be frustrating. And it's not that we don't know how many rows are there now (after using `free_results()`). We still have this information, so why not make it available? Especially if there is a simple workaround - like in this case.

I would expect Result->freeResult to also clear out any internal variables containing all those records you fetched so that you can save memory. That CI3 does not is surprising. However, to move forward with this PR, I will push another commit that mimics CI3's behavior.

>     1. if debug mode is enabled (`$this->DBDebug`) then we should throw an exception. Regardless of the status for debug mode we should always use `log_message('error', 'error message')` for logging the error

thanks for that detail!

>     2. if debug mode is not enabled, then we should return `0` because I feel like returning numeric value makes more sense here. Internally when`$numRows` is `null`, he knows that `getNumRows()` were never called.

Would it be OK with you for `Result::getNumRows()` to return `null` or `false` if `resultID` is a boolean or if it's a non-scrolling type of db result? This seems like a fairly innocuous way to signal an error condition and these become zero when treated as integers. I.e., `intval(false)` and `intval(null)` return zero.

Reply:
> Thank you very much for the code example. I've never seen CodeIgniter code that would extract a query's results, free the result object, and then refer to the freed query result to get a record count when you already have a variable containing those results.

I'm not saying that these examples represent coding standards of high-quality but they should give us a feeling of how people can use this method.

> IMHO, a proper freeResult method would also clear any resultArray or resultObject arrays inside the query object. On the other hand, migration concerns and backwards compatibility matter. One can always do unset($query), right?

Backward compatibility is a pain in the ass sometimes but there is no other way.

> Would it be OK with you for Result::getNumRows() to return null or false if resultID is a boolean or if it's a non-scrolling type of db result? This seems like a fairly innocuous way to signal an error condition and these become zero when treated as integers. I.e., intval(false) and intval(null) return zero.

I would prefer to `getNumRows()` return `int` value only. We should not promote using this method as an indicator of errors for the query. It should be checked elsewhere. So, I would say that returning a numeric value would be exactly what we expect from this method.